### ISSUE TITLE
`LiteralString` type is not enforced in invariant context

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5999]
The special form `LiteralString` is a subtype of `str`, and all literal string values are subtypes of `LiteralString`. When `LiteralString` is used in an invariant context (e.g. `list[LiteralString]`), it should be considered incompatible with either `str` or a specific literal str.

Currently, the following code generates no errors, but it should produce two errors.

```python
def func(a: list[LiteralString], b: list[Literal["a"]]):
    # This should generate an error because of invariance rules.
    v1: list[str] = a

    # This should generate an error because of invariance rules.
    v2: list[LiteralString] = b
```

**Comment by mikeshardmind** [comment]
I'm not completely convinced that either of these cases are bugs, but inverting the LHS and RHS, they would be.

The problem in this case is that these are essentially refinement types. This isn't a normal subtyping relationship. All LiteralStrings are perfect drop in replacements for strings at runtime, but the reverse is not true

**Comment by erictraut** [comment]
The term "refinement type" is new to me. I don't think it has been used in any Python type documentation or PEPs previously (including [PEP 586](https://peps.python.org/pep-0586/) which introduced `Literal` and [PEP 675](https://peps.python.org/pep-0675/) which introduced `LiteralString`). Are you suggesting that we should introduce a new concept into the Python type system that doesn't exist today — and introduce exceptions whereby these types don't follow the same subtyping rules as other types?

Logically, it makes sense to me that `Literal[False]` and `Literal[True]` are proper subtypes of `bool`. Likewise, it makes sense that `Literal["a"]` is a proper subtype of `LiteralString`, which represents the union of all possible literal strings. And `LiteralString` is a proper subtype of `str`. That seems consistent with type theory rules. Treating literal types differently from other types in the type system would 

I'm going to stick with my current understanding of how literal types work in the Python type system. If some future formal type specification clarifies that these should work differently, then we can make the appropriate changes at that time. Such a change would impact more than just `LiteralString`, and it would require changes to all Python type checkers. In the meantime, I'm going to maintain consistency — both internally within pyright and across type checkers — by sticking to my current understanding.

**Comment by mikeshardmind** [comment]
https://en.wikipedia.org/wiki/Refinement_type is somewhat useful here for explaining, the definition here is accurate. 

Namely, a type specification: `Literal["x"]` is the set of all string literals that are "x"

It may be more intuitive to consider a numeric case. If we had `Literal[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]`, any value of this is clearly exactly an int at runtime, and safe as a complete substitute for one, but these are also known to be safe to pack into a specific bitsize, whereas this isn't true of all ints

**Comment by erictraut** [comment]
Hmm, I'm not sure I follow. Wouldn't that mean the following would type check without error? Isn't this a problem?

```python
from typing import Literal

def append_b(x: list[str]):
    x.append("b")

def func1(x: list[Literal["a"]]):
    append_b(x)
    assert all([v == "a" for v in x])

func1(["a"])
```

**Comment by mikeshardmind** [comment]
That's different from your initial example and falls under what I said about swapping the LHS and RHS, refinement types have a bit more nuance in where they can be soundly substituted and that's even something I mentioned in one of these recent discussions about how intersections have gotten complex due to lack of clarity in places in the type system.

Wtih basic classes as types, we have nominal subtyping
With protocols, we have structural subtyping
With Literals, we have refinement types

and the goal is to ensure we reconcile the behavior of all of them.
___

I think it's "fine" but not ideal to treat this as subtyping temporarily until this can be cleared up at a specification level. This is more restrictive than it needs to be, your most recent example should fail, but both of your initial examples should be fine.

**Comment by DiscordLiz** [comment]
@mikeshardmind Agreed on specification and more formal language helping, but in this case, it's 100% fine to just treat a refinement type as a subtype. If you step back for a moment, you'll notice the handling the variance as you would for mutability already handles where it is and isn't safe. This matches with the set-theoretic definitions I hope are adopted long term.

@erictraut I'd like to apologize about prior interactions while I'm here. I let my frustration with the status quo inappropriately effect how I was participating.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.329, which I just published. It will also be included in a future release of pylance.

**Comment by zwade** [comment]
Hi @erictraut, sorry to comment on an already closed issue, but I was wondering if there was an followup to @mikeshardmind's comment above

> However, what about a frozen dataclass? It's perfectly fine to put Literal[1] in place of a field in one of those expecting an int.

I have this particular use case and am wondering how to proceed. Is there a way to mark a property as read-only so that the subtyping context is treated as covariant instead of invariant?

Thanks!

**Comment by erictraut** [comment]
@zwade, the latest version of pyright (1.1.329) treats attributes within frozen dataclasses as immutable. See [this issue](https://github.com/microsoft/pyright/issues/6029) and [this issue](https://github.com/microsoft/pyright/issues/5993). If your use case isn't addressed by this, please provide more details along with a self-contained minimal code sample.
